src/Storages/Cache/ExternalDataSourceCache.cpp

```diff
@@ -0,0 +1,244 @@
+}
+
+std::unique_ptr<ReadBuffer> RemoteReadBuffer::create(ContextPtr context, IRemoteFileMetadataPtr remote_file_metadata, std::unique_ptr<ReadBuffer> read_buffer, size_t buff_size)
+{
+    auto remote_path = remote_file_metadata->remote_path;
+    auto remote_read_buffer = std::make_unique<RemoteReadBuffer>(buff_size);
+    ErrorCodes::ErrorCode error;
+
+    std::tie(remote_read_buffer->file_cache_controller, read_buffer, error) = ExternalDataSourceCache::instance().createReader(context, remote_file_metadata, read_buffer);
+    if (remote_read_buffer->file_cache_controller == nullptr)
+    {
+        return read_buffer;
+    }
+    else
+    {
+        remote_read_buffer->file_buffer = remote_read_buffer->file_cache_controller->allocFile();
+        if (!remote_read_buffer->file_buffer)
+            throw Exception(ErrorCodes::LOGICAL_ERROR, "Create file readbuffer failed. {}",
+                    remote_read_buffer->file_cache_controller->getLocalPath().string());
+    }
+    remote_read_buffer->remote_file_size = remote_file_metadata->file_size;
+    return remote_read_buffer;
+}
+
+bool RemoteReadBuffer::nextImpl()
+{
+    auto start_offset = file_buffer->getPosition();
+    auto end_offset = start_offset + file_buffer->internalBuffer().size();
+    file_cache_controller->waitMoreData(start_offset, end_offset);
+
+    auto status = file_buffer->next();
+    if (status)
+    {
+        BufferBase::set(file_buffer->buffer().begin(),
+                file_buffer->buffer().size(),
+                file_buffer->offset());
+        ProfileEvents::increment(ProfileEvents::ExternalDataSourceLocalCacheReadBytes, file_buffer->available());
+    }
+    return status;
+}
+
+off_t RemoteReadBuffer::seek(off_t offset, int whence)
+{
+    if (!file_buffer)
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "Cannot call seek() in this buffer. It's a bug!");
+    /*
+     * Need to wait here. For example, the current file has been download at position X, but here we try to seek to
+     * postition Y (Y > X), it would fail.
+     */
+    file_cache_controller->waitMoreData(offset, offset + file_buffer->internalBuffer().size());
+    auto ret = file_buffer->seek(offset, whence);
+    BufferBase::set(file_buffer->buffer().begin(),
+            file_buffer->buffer().size(),
+            file_buffer->offset());
+    return ret;
+}
+
+off_t RemoteReadBuffer::getPosition()
+{
+    return file_buffer->getPosition();
+}
+
+ExternalDataSourceCache::ExternalDataSourceCache() = default;
+
+ExternalDataSourceCache::~ExternalDataSourceCache()
+{
+    recover_task_holder->deactivate();
+}
+
+ExternalDataSourceCache & ExternalDataSourceCache::instance()
+{
+    static ExternalDataSourceCache instance;
+    return instance;
+}
+
+void ExternalDataSourceCache::recoverCachedFilesMetadata(
+    const fs::path & current_path,
+    size_t current_depth,
+    size_t max_depth)
+{
+    if (current_depth >= max_depth)
```

Comment:
May be remove this recursive recovery? because in fact we always need to recover only last level.

Reply:
done